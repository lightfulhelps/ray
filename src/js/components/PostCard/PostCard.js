// @flow
import * as React from 'react';
import classNames from 'classnames';
import Dotdotdot from 'react-dotdotdot';
import formatDate from 'date-fns/format';
import { Card, Avatar, Badge, Icon, PostMedia, URLMetaPreview } from '../../';

type InspirationActionType = {
  activeColor?: string,
  color?: string,
  icon: string,
  isActive?: boolean,
  onClick?: () => void,
};

type MediaType = {
  type: string,
  url: string,
};

type PostType = {
  campaign?: {
    color: string,
    name: string,
  },
  content: string,
  date?: Date | number | string,
  media?: MediaType[],
  socialIdentity?: {
    avatar?: string,
    displayName?: string,
    id: string,
    username?: string,
  },
  socialProvider?: string,
  title: string,
};

type Props = {
  className?: string,
  dateFormat?: string,
  inspirationActions?: InspirationActionType[],
  isDraft?: boolean,
  isInvalid?: boolean,
  metaPreview?: {
    description?: string,
    image?: string,
    title: string,
    url: string,
  },
  post: PostType,
};

const PostCard = ({
  className,
  dateFormat = 'D MMM [at] HH:mm',
  inspirationActions,
  isDraft,
  isInvalid,
  metaPreview,
  post,
}: Props) => {
  const blockClass = 'post-card';
  const classes = classNames(
    className,
    blockClass,
    { [`${blockClass}--draft`]: isDraft },
    { [`${blockClass}--invalid`]: isInvalid }
  );

  return (
    <Card className={classes}>
      <div className={`${blockClass}__header`}>
        {post.socialIdentity && (
          <div style={{ width: '35px', height: '35px' }}>
            <Avatar url={post.socialIdentity.avatar} provider={post.socialProvider} />
          </div>
        )}
        <div className="mx-1">
          <h1 className={`${blockClass}__title`}>{post.title}</h1>
          {post.date && (
            <div className={`${blockClass}__date`}>{formatDate(post.date, dateFormat)}</div>
          )}
          {post.campaign && (
            <Badge className="campaign-tag" color={post.campaign.color}>
              {post.campaign.name}
            </Badge>
          )}
        </div>
      </div>
      <Dotdotdot className={`${blockClass}__content`} clamp={5}>
        {post.content}
      </Dotdotdot>
      {post.media && (
        <div className={`${blockClass}__media`}>
          <PostMedia media={post.media} />
        </div>
      )}
      {metaPreview && (
        <div className={`${blockClass}__media border-top`}>
          <URLMetaPreview {...metaPreview} />
        </div>
      )}
      {inspirationActions &&
        inspirationActions.length > 0 && (
          <div className={`${blockClass}__inspiration-actions`}>
            {inspirationActions.map((action, i) => (
              <Icon
                key={i}
                name={action.icon}
                className="cursor-pointer"
                color={action.isActive ? action.activeColor : action.color}
                hoverColor={action.isActive ? action.activeColor : '#343a40'}
                onClick={action.onClick}
              />
            ))}
          </div>
        )}
    </Card>
  );
};

export default PostCard;
