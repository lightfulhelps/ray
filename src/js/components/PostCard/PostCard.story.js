import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import { Container, Row, PostCard } from '../../';

import '../../../scss/ray.scss';

const stories = storiesOf('PostCard', module);

stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <Container>
    <h1>PostCard.</h1>
    <Row>
      <div className="col-md-6 col-lg-4 mb-2">
        <PostCard
          content={text(
            'Content',
            'Buttle UK helped more than 3,000 vulnerable families buy beds for their children last year. It fears thousands more across the UK may lack a bed of their own, leading to problems concentrating in school. The government said its welfare reforms were "supporting those who need it most".'
          )}
          isDraft={boolean('Draft', false)}
          isInvalid={boolean('Invalid', false)}
        />
      </div>
    </Row>
  </Container>
));
