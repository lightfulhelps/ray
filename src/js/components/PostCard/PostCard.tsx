import React, { useState } from 'react';
import classNames from 'classnames';
import HTMLEllipsis from 'react-lines-ellipsis/lib/html';
import responsiveHOC from 'react-lines-ellipsis/lib/responsiveHOC';
import { format as formatDate } from 'date-fns';
import { Card, Avatar, PostMedia, URLMetaPreview, Button, Tag, Icon } from '../..';
import { IconNameType } from '../Icon/icons';
import { Props as ButtonProps } from '../Button/Button';
import Alert from '../Alert/Alert';

export type MediaType = {
  id: string;
  type: string;
  url: string;
};

export type PostType = {
  campaign?: {
    color: string;
    name: string;
  };
  content: string;
  date?: Date | number | string;
  id: string;
  media?: MediaType[];
  metrics?: {
    icon?: IconNameType;
    key: string;
    value: number;
  }[];
  socialIdentity?: {
    avatar?: string;
    displayName?: string;
    id: string;
    provider: 'facebook' | 'twitter' | 'linkedin';
    username?: string;
  };
  state: string;
  tags: string[];
};

export type Props = {
  actions?: ButtonProps[];
  addLabel?: any;
  className?: string;
  dateFormat?: string;
  errors?: string[];
  isDraft?: boolean;
  isImported?: boolean;
  isInvalid?: boolean;
  metaPreview?: {
    description?: string;
    image?: string;
    title: string;
    url: string;
  };
  notesAction?: () => void;
  notesCount?: number;
  post: PostType;
  [key: string]: any;
};

const ResponsiveHTMLEllipsis = responsiveHOC()(HTMLEllipsis);

const PostCard: React.FC<Props> = props => {
  const [isTruncated, setIsTruncated] = useState(true);

  const handleToggleTruncate = () => {
    setIsTruncated(!isTruncated);
  };

  const {
    actions = [],
    addLabel,
    className,
    dateFormat = "HH:mm 'on' eeee, d MMMM",
    errors,
    isDraft,
    isImported,
    isInvalid,
    metaPreview,
    notesAction,
    notesCount = 0,
    post,
    ...other
  } = props;
  const classes = classNames(
    className,
    'post-card shadow rounded overflow-hidden',
    { 'post-card--draft': isDraft },
    { 'post-card--invalid': isInvalid }
  );

  if (!post) return null;

  const showMetaPreview =
    (!post.media || (post.media && post.media.length === 0)) && metaPreview && metaPreview.url;
  const showMediaEmpty =
    post.media && post.media.length === 0 && (!metaPreview || (metaPreview && !metaPreview.url));
  const hasPostMetrics = post.metrics && post.metrics.length > 0;

  const defaultLineClamp = 3;
  const allLinesClamp = 9999;

  let borderColor;

  switch (post.state) {
    case 'scheduled':
      borderColor = 'gradient-primary-x';
      break;
    case 'published':
      borderColor = 'success';
      break;
    case 'review':
      borderColor = 'warning';
      break;
    default:
      borderColor = '';
      break;
  }

  return (
    <Card {...other} className={classes}>
      <div className={`bg-${borderColor} rounded-top mb-1`} style={{ height: '5px' }} />
      {errors && errors.length > 0 && (
        <div className="mb-1">
          {errors.map((error, i) => (
            <Alert
              key={i}
              className={classNames('mx-3', { 'mb-1': i < errors.length - 1 })}
              theme="danger"
            >
              {error}
            </Alert>
          ))}
        </div>
      )}
      {isImported && (
        <div className="post-card__imported px-1 py-half rounded-sm text-sm alert-info fw-normal">
          This post was imported from outside of Lightful. Link clicks are not tracked.
        </div>
      )}
      <div className="d-flex flex-column flex-md-row justify-content-between px-3 pb-3 pt-1 border-bottom">
        <div className="flex-fill">
          <div className="d-flex mb-3">
            {post.socialIdentity && (
              <Avatar
                className="flex-shrink-0"
                url={post.socialIdentity.avatar}
                provider={post.socialIdentity.provider}
                style={{ width: '45px', height: '45px' }}
              />
            )}
            <div className="ms-1">
              <div className="h4 mb-0" data-test-id="post-card-date">
                {post.date
                  ? `${post.state === 'published' ? 'Published' : 'Scheduled for'} ${formatDate(
                      new Date(post.date),
                      dateFormat
                    )}`
                  : 'Unscheduled'}
              </div>
              {post.socialIdentity && (
                <div className="text-sm">{post.socialIdentity.displayName}</div>
              )}
            </div>
          </div>
          <div
            className="mb-3 post-card__content"
            data-test-id="post-card-content"
            onClick={handleToggleTruncate}
          >
            <ResponsiveHTMLEllipsis
              unsafeHTML={post.content.replace(/\n/g, '<br />')}
              maxLine={isTruncated ? defaultLineClamp : allLinesClamp}
              ellipsisHTML='<span class="text-underline cursor-pointer fw-bold">See more</span>'
              basedOn="words"
            />
          </div>
          <div className="d-flex flex-wrap">
            {post.campaign && (
              <Tag className="d-flex align-items-center mb-1 me-1 badge-pill" icon="storyBuilder">
                Story: {post.campaign.name}
              </Tag>
            )}
            {post.tags &&
              post.tags.length > 0 &&
              post.tags.map((tag, i) => (
                <Tag
                  className={`me-1 mb-1 ${i === 0 ? '' : ''}`}
                  isOutline
                  key={i}
                  theme="gray-600"
                  icon="tag"
                >
                  {tag}
                </Tag>
              ))}
            {addLabel && (
              <Button
                isOutline
                icon="add"
                className="me-1 mb-1 rounded-pill"
                data-test-id="post-card-add-label"
                {...addLabel}
              />
            )}
          </div>
        </div>
        <div
          className={classNames('post-card__media-wrap flex-shrink-0', {
            'h-100': showMetaPreview,
          })}
        >
          {post.media && post.media.length > 0 && <PostMedia media={post.media} />}
          {showMetaPreview && <URLMetaPreview {...metaPreview} className="shadow" isVertical />}
          {showMediaEmpty && <div className="post-media--empty">No media</div>}
        </div>
      </div>
      <div
        className={classNames(
          'py-2 px-3 d-flex flex-column flex-md-row align-items-center',
          notesAction || hasPostMetrics ? 'justify-content-between' : 'justify-content-end'
        )}
      >
        {(notesAction || hasPostMetrics) && (
          <div className="d-flex align-items-center flex-wrap mt-1 mt-md-0 mb-3 mb-md-0">
            {notesAction ? (
              <div
                className="d-flex align-items-center cursor-pointer"
                data-test-id="post-card-notes"
                onClick={notesAction}
              >
                <Icon className="me-half" name="comment" />{' '}
                <span className="text-underline text-sm" style={{ lineHeight: 1 }}>
                  View notes ({notesCount.toString()})
                </span>
              </div>
            ) : (
              post.metrics &&
              post.metrics.length > 0 &&
              post.metrics.map(metric => (
                <div
                  className="d-flex align-items-center text-sm me-1"
                  data-test-id="post-card-metric"
                  key={metric.key}
                >
                  {metric.icon && (
                    <Icon className="me-half" isActive name={metric.icon} size={20} />
                  )}{' '}
                  {metric.value} {metric.key}
                </div>
              ))
            )}
          </div>
        )}
        <div className="d-flex flex-wrap">
          {actions.length > 0 &&
            actions.map((action, i) => (
              <Button
                key={i}
                isOutline
                className="me-3 mb-1 me-md-0 ms-md-3 mb-lg-0"
                data-test-id="post-card-action"
                {...action}
              />
            ))}
        </div>
      </div>
    </Card>
  );
};

export default PostCard;
