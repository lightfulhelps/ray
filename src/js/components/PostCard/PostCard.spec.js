import React from 'react';
import { shallow } from 'enzyme';
import PostCard from './PostCard';

const setup = (overrides = {}) => {
  const props = {
    ...overrides,
  };
  const wrapper = shallow(<PostCard {...props} />);

  return {
    wrapper,
    props,
  };
};

describe('<PostCard />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle the className prop', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('post-card')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });
});
