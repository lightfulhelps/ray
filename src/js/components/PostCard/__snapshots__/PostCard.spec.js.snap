// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostCard /> should render 1`] = `
<Card
  className="post-card shadow"
>
  <div
    className="bg-info rounded-top-sm"
    style={
      Object {
        "height": "3px",
      }
    }
  />
  <div
    className="d-flex flex-column flex-md-row justify-content-between p-2 border-bottom"
  >
    <div
      className="flex-fill"
    >
      <div
        className="d-flex mb-1"
      >
        <Avatar
          className="flex-shrink-0"
          provider="facebook"
          style={
            Object {
              "height": "40px",
              "width": "40px",
            }
          }
          url="http://foo.com/avatar.jpg"
        />
        <div
          className="ml-1"
        >
          <div
            className="h6 mb-0"
            data-test-id="post-card-date"
          >
            Scheduled for 14:34 on Wednesday, 22 August
          </div>
          <div
            className="text-sm"
          >
            Myah Graham
          </div>
        </div>
      </div>
      <div
        className="text-sm mb-1"
        data-test-id="post-card-content"
      >
        <Responsive(HTMLEllipsis)
          basedOn="words"
          ellipsis="..."
          innerRef={[Function]}
          maxLine={3}
          unsafeHTML="Buttle UK helped more than 3,000 vulnerable families buy beds for their children last year. It fears thousands more across the UK may lack a bed of their own, leading to problems concentrating in school. The government said its welfare reforms were \\"supporting those who need it most\\"."
        />
        <div
          className="d-flex justify-content-end"
          onClick={[Function]}
        >
          <div
            className="cursor-pointer text-underline"
          >
            See more
          </div>
        </div>
      </div>
      <div
        className="d-flex flex-wrap"
      >
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="0"
          theme="gray-600"
        >
          charity
        </Tag>
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="1"
          theme="gray-600"
        >
          social for good
        </Tag>
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="2"
          theme="gray-600"
        >
          campaigns
        </Tag>
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="3"
          theme="gray-600"
        >
          lightful
        </Tag>
      </div>
    </div>
    <div
      className="post-card__media-wrap flex-shrink-0 overflow-hidden"
    >
      <PostMedia
        media={
          Array [
            Object {
              "id": "af712270-b03e-4b1c-98d2-f744f807df2f",
              "type": "image",
              "url": "http://lorempixel.com/640/480",
            },
          ]
        }
      />
    </div>
  </div>
  <div
    className="py-1 px-2 d-flex flex-column flex-md-row justify-content-between align-items-center"
  >
    <div
      className="d-flex align-items-center flex-wrap mb-1 mb-md-0"
    >
      <div
        className="d-flex align-items-center text-sm mr-1"
        key="likes"
      >
        <Icon
          className="mr-half"
          isActive={true}
          name="like"
          size={20}
        />
         
        10
         
        likes
      </div>
      <div
        className="d-flex align-items-center text-sm mr-1"
        key="shares"
      >
        <Icon
          className="mr-half"
          isActive={true}
          name="share"
          size={20}
        />
         
        5
         
        shares
      </div>
    </div>
    <div
      className="d-flex flex-wrap"
    />
  </div>
</Card>
`;

exports[`<PostCard /> should render with errors 1`] = `
<Card
  className="post-card shadow"
>
  <div
    className="bg-info rounded-top-sm"
    style={
      Object {
        "height": "3px",
      }
    }
  />
  <div
    className="mt-1"
  >
    <div
      className="text-xs font-weight-bold mb-1 mx-2 bg-danger py-half px-1 rounded-sm text-white"
      key="0"
    >
      foo
    </div>
    <div
      className="text-xs font-weight-bold mb-1 mx-2 bg-danger py-half px-1 rounded-sm text-white"
      key="1"
    >
      bar
    </div>
  </div>
  <div
    className="d-flex flex-column flex-md-row justify-content-between p-2 border-bottom"
  >
    <div
      className="flex-fill"
    >
      <div
        className="d-flex mb-1"
      >
        <Avatar
          className="flex-shrink-0"
          provider="facebook"
          style={
            Object {
              "height": "40px",
              "width": "40px",
            }
          }
          url="http://foo.com/avatar.jpg"
        />
        <div
          className="ml-1"
        >
          <div
            className="h6 mb-0"
            data-test-id="post-card-date"
          >
            Scheduled for 14:34 on Wednesday, 22 August
          </div>
          <div
            className="text-sm"
          >
            Myah Graham
          </div>
        </div>
      </div>
      <div
        className="text-sm mb-1"
        data-test-id="post-card-content"
      >
        <Responsive(HTMLEllipsis)
          basedOn="words"
          ellipsis="..."
          innerRef={[Function]}
          maxLine={3}
          unsafeHTML="Buttle UK helped more than 3,000 vulnerable families buy beds for their children last year. It fears thousands more across the UK may lack a bed of their own, leading to problems concentrating in school. The government said its welfare reforms were \\"supporting those who need it most\\"."
        />
        <div
          className="d-flex justify-content-end"
          onClick={[Function]}
        >
          <div
            className="cursor-pointer text-underline"
          >
            See more
          </div>
        </div>
      </div>
      <div
        className="d-flex flex-wrap"
      >
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="0"
          theme="gray-600"
        >
          charity
        </Tag>
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="1"
          theme="gray-600"
        >
          social for good
        </Tag>
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="2"
          theme="gray-600"
        >
          campaigns
        </Tag>
        <Tag
          className="text-xs mr-half mb-1 "
          isOutline={true}
          key="3"
          theme="gray-600"
        >
          lightful
        </Tag>
      </div>
    </div>
    <div
      className="post-card__media-wrap flex-shrink-0 overflow-hidden"
    >
      <PostMedia
        media={
          Array [
            Object {
              "id": "af712270-b03e-4b1c-98d2-f744f807df2f",
              "type": "image",
              "url": "http://lorempixel.com/640/480",
            },
          ]
        }
      />
    </div>
  </div>
  <div
    className="py-1 px-2 d-flex flex-column flex-md-row justify-content-between align-items-center"
  >
    <div
      className="d-flex align-items-center flex-wrap mb-1 mb-md-0"
    >
      <div
        className="d-flex align-items-center text-sm mr-1"
        key="likes"
      >
        <Icon
          className="mr-half"
          isActive={true}
          name="like"
          size={20}
        />
         
        10
         
        likes
      </div>
      <div
        className="d-flex align-items-center text-sm mr-1"
        key="shares"
      >
        <Icon
          className="mr-half"
          isActive={true}
          name="share"
          size={20}
        />
         
        5
         
        shares
      </div>
    </div>
    <div
      className="d-flex flex-wrap"
    />
  </div>
</Card>
`;
